src/nci_util.c

```diff
@@ -849,6 +849,105 @@ nci_parse_intf_activated_ntf(
+                copy->poll_b.prot_info.bytes = dest;
+                size += G_ALIGN8(poll_b->prot_info.size);
+            }
+            return size;
+        case NCI_MODE_ACTIVE_LISTEN_F:
+        case NCI_MODE_PASSIVE_LISTEN_F:
+            listen_f = &param->listen_f;
+            GASSERT(!listen_f->nfcid2.bytes || listen_f->nfcid2.size);
+            GASSERT(!listen_f->nfcid2.size || listen_f->nfcid2.bytes);
+            memcpy(copy, param, sizeof(NciModeParam));
+            if (listen_f->nfcid2.bytes)
+            {
+                guint8* dest = (guint8*)copy + size;
+                memcpy(dest, listen_f->nfcid2.bytes, listen_f->nfcid2.size);
+                copy->listen_f.nfcid2.bytes = dest;
+                size += G_ALIGN8(listen_f->nfcid2.size);
+            }
+            return size;
+        case NCI_MODE_PASSIVE_POLL_15693:
+        case NCI_MODE_PASSIVE_LISTEN_15693:
+            break;
+        case NCI_MODE_PASSIVE_LISTEN_A:     /* fallthrough */
+        case NCI_MODE_PASSIVE_LISTEN_B:     /* fallthrough */
+        case NCI_MODE_ACTIVE_LISTEN_A:
+            /* NCI 1.0 defines no parameters for A/B Listen modes */
+            return 0;
+        }
+        GDEBUG("Unhandled activation mode 0x%02x", mode);
+        return 0;
+    } else {
+        return 0;
+    }
+}
+
+static gsize calculate_mode_param_fit_size(
+    const NciModeParam* param,
+    NCI_MODE mode)
+{
+    gsize size = G_ALIGN8(sizeof(NciModeParam));
+    if (G_LIKELY(param)) {
+        const NciModeParamPollB* poll_b = NULL;
+        const NciModeParamListenF* listen_f = NULL;
+
+        switch (mode) {
+        case NCI_MODE_PASSIVE_POLL_B:
+            poll_b = &param->poll_b;
+            GASSERT(!poll_b->prot_info.bytes || poll_b->prot_info.size);
+            GASSERT(!poll_b->prot_info.size || poll_b->prot_info.bytes);
+            size += G_ALIGN8(poll_b->prot_info.size);
+            break;
+        case NCI_MODE_ACTIVE_LISTEN_F:
+        case NCI_MODE_PASSIVE_LISTEN_F:
+            listen_f = &param->listen_f;
+            GASSERT(!listen_f->nfcid2.bytes || listen_f->nfcid2.size);
+            GASSERT(!listen_f->nfcid2.size || listen_f->nfcid2.bytes);
+            size += G_ALIGN8(listen_f->nfcid2.size);
+            break;
+        case NCI_MODE_ACTIVE_POLL_A:        /* fallthrough */
+        case NCI_MODE_PASSIVE_POLL_A:       /* fallthrough */
+        case NCI_MODE_ACTIVE_POLL_F:        /* fallthrough */
+        case NCI_MODE_PASSIVE_POLL_F:       /* fallthrough */
+        case NCI_MODE_PASSIVE_POLL_15693:   /* fallthrough */
+        case NCI_MODE_PASSIVE_LISTEN_15693: /* fallthrough */
+        case NCI_MODE_PASSIVE_LISTEN_A:     /* fallthrough */
+        case NCI_MODE_PASSIVE_LISTEN_B:     /* fallthrough */
+        case NCI_MODE_ACTIVE_LISTEN_A:      /* fallthrough */
+            /* NCI 1.0 defines no parameters for A/B Listen modes */
+            break;
```

Comment:
There's some inconsistency here. The copying function does one thing (doesn't copy anything and returns zero) and the sizing function another (returns non-zero) for unsupported modes and modes that don't have parameters.